lib/charms/kratos_external_idp_integrator/v0/kratos_external_provider.py

```diff
@@ -0,0 +1,347 @@
+    providers = [
+        "google",
+        "facebook",
+        "github",
+        "gitlab",
+        "slack",
+        "spotify",
+        "discord",
+        "twitch",
+        "netid",
+        "yander",
+        "vk",
+        "dingtalk",
+    ]
+
+    @classmethod
+    def _parse_provider_config(cls, config):
+        return {
+            "client_secret": config["client_secret"],
+        }
+
+
+class MicrosoftConfigHandler(SocialConfigHandler):
+    """The class for parsing a 'microsoft' provider's config."""
+
+    mandatory_fields = SocialConfigHandler.mandatory_fields | {
+        "microsoft_tenant_id",
+    }
+    providers = ["microsoft"]
+
+    @classmethod
+    def _parse_provider_config(cls, config):
+        return {
+            "client_secret": config["client_secret"],
+            "tenant_id": config["microsoft_tenant_id"],
+        }
+
+
+class AppleConfigHandler(BaseProviderConfigHandler):
+    """The class for parsing an 'apple' provider's config."""
+
+    mandatory_fields = BaseProviderConfigHandler.mandatory_fields | {
+        "apple_team_id",
+        "apple_private_key_id",
+        "apple_private_key",
+    }
+    _secret_fields = ["private_key"]
+    providers = ["apple"]
+
+    @classmethod
+    def _parse_provider_config(cls, config):
+        return {
+            "team_id": config["apple_team_id"],
+            "private_key_id": config["apple_private_key_id"],
+            "private_key": config["apple_private_key"],
+        }
+
+
+_config_handlers = [
+    GenericConfigHandler,
+    SocialConfigHandler,
+    MicrosoftConfigHandler,
+    AppleConfigHandler,
+]
+allowed_providers = {
+    provider: handler for handler in _config_handlers for provider in handler.providers
+}
+
+
+class RedirectURIChangedEvent(EventBase):
+    """Event to notify the charm that the redirect_uri changed."""
+
+    def __init__(self, handle, redirect_uri):
+        super().__init__(handle)
+        self.redirect_uri = redirect_uri
+
+    def snapshot(self):
+        """Save redirect_uri."""
+        return {"redirect_uri": self.redirect_uri}
+
+    def restore(self, snapshot):
+        """Restore redirect_uri."""
+        self.redirect_uri = snapshot["redirect_uri"]
+
+
+class InvalidClientConfigEvent(EventBase):
+    """Event to notify the charm that the requirer's databag is invalid."""
+
+    def __init__(self, handle, error):
+        super().__init__(handle)
+        self.error = error
+
+    def snapshot(self):
+        """Save error."""
+        return {"error": self.error}
+
+    def restore(self, snapshot):
+        """Restore error."""
+        self.error = snapshot["error"]
+
+
+class ExternalIdpProviderEvents(ObjectEvents):
+    """Event descriptor for events raised by `ExternalIdpProvider`."""
+
+    redirect_uri_changed = EventSource(RedirectURIChangedEvent)
+    invalid_client_config = EventSource(InvalidClientConfigEvent)
+
+
+class ExternalIdpProvider(Object):
```

Comment:
The provider/requirer convention is not entirely clear to me, but I think this should be the Provider side (as it provides the client credentials and Kratos uses them). That's why I named it like this, I have not pushed the requirer side yet. If you think the roles should be reversed let me know.

Reply:
I was thinking that this charm requires Kratos, so it was more of a requirer. I was further confused because the docstring instructs charm to set a require relation to use the provider class. I could see your point though. I am good as long as the provides side of the relation uses the provider class

Reply:
I forgot to change that, sorry for the confusion.